### PR TITLE
fix(store/cache): force close Accessor if not closed within timeout

### PR DESCRIPTION
**Pull request by walldiss**
- Force close Accessor if not closed within timeout

### CODE REVIEW

**Review by Wondertan — APPROVED** [review]

File: share/eds/cache/accessor_cache.go [hunk 1]
```diff
@@ -244,17 +248,31 @@ func newRefCloser(abs *accessorWithBlockstore) (*refCloser, error) {
...
-		},
-	}, nil
+		removeRef:              abs.removeRef,
+	}
+	// Set finalizer to ensure that accessor is closed when refCloser is garbage collected.
+	// We expect that refCloser will be closed explicitly by the caller. If it is not closed,
+	// we log an error.
+	runtime.SetFinalizer(rf, func(rf *refCloser) {
+		if rf.close() {
+			log.Errorf("refCloser for accessor was garbage collected before Close was called")
```
Comment (renaynay) [inline share/eds/cache/accessor_cache.go:260]:
```suggestion
			log.Error("refCloser for accessor was garbage collected before Close was called")
```

**Review by renaynay — APPROVED** [review]

File: share/eds/cache/accessor_cache.go [hunk 2]
```diff
@@ -244,17 +248,31 @@ func newRefCloser(abs *accessorWithBlockstore) (*refCloser, error) {
...
-		closeFn: func() {
-			closeOnce.Do(abs.removeRef)
-		},
-	}, nil
+		removeRef:              abs.removeRef,
+	}
+	// Set finalizer to ensure that accessor is closed when refCloser is garbage collected.
+	// We expect that refCloser will be closed explicitly by the caller. If it is not closed,
+	// we log an error.
+	runtime.SetFinalizer(rf, func(rf *refCloser) {
```
Comment (cristaloleg) [inline share/eds/cache/accessor_cache.go:258]: How many such invocations we will have? Approx.

Comment (walldiss) [reply share/eds/cache/accessor_cache.go:258]: few(1-10) per sec at most

Comment (cristaloleg) [reply share/eds/cache/accessor_cache.go:258]: upto 100 per **second** ??

Comment (cristaloleg) [reply share/eds/cache/accessor_cache.go:258]: Why I'm asking: `runtime.SetFinalizer` behaviour is easy to misuse, see docs (https://pkg.go.dev/runtime#SetFinalizer) & proposal to deprecated it (https://github.com/golang/go/issues/67535)

Adding up to 100/sec looks possible but better to bring attention to this line and solution in general, my 2c.

Comment (walldiss) [reply share/eds/cache/accessor_cache.go:258]: This is one of 2 safety solutions we use for preventing hanging ref from blocking Accessor eviction from cache. 
- Finaliser closing forgotten ref.
- Force closing forgotten refs on timeout.

Given the fact, it is safety only mechanics and not main case. And assuming the only client of cache library is us, we can get away without Finaliser. Considering implications and concerns about Finaliser I'm inclined to remove it in favour of remaining forced closing only. 

@Wondertan thoughts?

Comment (cristaloleg) [reply share/eds/cache/accessor_cache.go]: Cool

**Review by Wondertan — APPROVED** [review]

**Review by vgonkivs — APPROVED** [review]

**Review by renaynay — APPROVED** [review]
Okay

**Pull request merged**